src/librustc_mir/transform/type_check.rs

```diff
@@ -0,0 +1,549 @@
+        let mut selcx = traits::SelectionContext::new(&self.infcx);
+        let cause = traits::ObligationCause::misc(self.last_span, 0);
+        let traits::Normalized { value: ty, obligations } =
+            traits::normalize(&mut selcx, cause, &ty);
+
+        debug!("normalize: ty={:?} obligations={:?}",
+               ty,
+               obligations);
+
+        let mut fulfill_cx = self.infcx.fulfillment_cx.borrow_mut();
+        for obligation in obligations {
+            fulfill_cx.register_predicate_obligation(&self.infcx, obligation);
+        }
+
+        match infer::drain_fulfillment_cx(&self.infcx, &mut fulfill_cx, &ty) {
+            Ok(ty) => ty,
+            Err(e) => {
+                span_mirbug_and_err!(self, parent, "trait fulfillment failed: {:?}", e)
+            }
+        }
+    }
+}
+
+pub struct TypeckMir<'a, 'tcx: 'a> {
+    infcx: &'a infer::InferCtxt<'a, 'tcx>,
+    last_span: Span
+}
+
+impl<'a, 'tcx> TypeckMir<'a, 'tcx> {
+    pub fn new(infcx: &'a infer::InferCtxt<'a, 'tcx>) -> Self {
+        TypeckMir {
+            infcx: infcx,
+            last_span: DUMMY_SP
+        }
+    }
+
+    fn tcx(&self) -> &'a ty::ctxt<'tcx> {
+        self.infcx.tcx
+    }
+
+    fn check_stmt(&mut self, mir: &Mir<'tcx>, stmt: &Statement<'tcx>) {
+        debug!("check_stmt: {:?}", stmt);
+        let tcx = self.tcx();
+        match stmt.kind {
+            StatementKind::Assign(ref lv, ref rv) => {
+                let lv_ty = mir.lvalue_ty(tcx, lv).to_ty(tcx);
+                let rv_ty = mir.rvalue_ty(tcx, rv);
+                if let Some(rv_ty) = rv_ty {
+                    if let Err(terr) = infer::can_mk_subty(self.infcx, rv_ty, lv_ty) {
+                        span_mirbug!(self, stmt, "bad assignment ({:?} = {:?}): {:?}",
+                                     lv_ty, rv_ty, terr);
+                    }
+                }
+
+                // FIXME: rvalue with undeterminable type - e.g. inline
+                // asm.
+            }
+        }
+    }
+
+    fn check_terminator(&self,
+                        mir: &Mir<'tcx>,
+                        term: &Terminator<'tcx>) {
+        debug!("check_terminator: {:?}", term);
+        let tcx = self.tcx();
+        match *term {
+            Terminator::Goto { .. } |
+            Terminator::Resume |
+            Terminator::Return |
+            Terminator::Drop { .. } => {}
+            Terminator::If { ref cond, .. } => {
+                let cond_ty = mir.operand_ty(tcx, cond);
+                match cond_ty.sty {
+                    ty::TyBool => {}
+                    _ => {
+                        span_mirbug!(self, term, "bad If ({:?}, not bool", cond_ty);
+                    }
+                }
+            }
+            Terminator::SwitchInt { ref discr, switch_ty, .. } => {
+                let discr_ty = mir.lvalue_ty(tcx, discr).to_ty(tcx);
+                if let Err(terr) = infer::can_mk_subty(self.infcx, discr_ty, switch_ty) {
```

Comment:
Also, the right way to use inference is not to call `can_mk_subty` but just `mk_subty`, since this is something that must be true (or else error). I don't think there are any inference variables floating around, so the difference is not large, but if you use `can_mk_subty`, then it will just check whether you *could* require that `discr_ty <: switch_ty`, versus adding the constraint that indeed `discr_ty <: swich_ty` ought to hold.